### PR TITLE
Jinja macro fixes, sync definition with documentation and implementation

### PR DESCRIPTION
**Pull request by maage**
#### Description:

This is continuation to #10599 and contains changes where there is macro modification.

- updated style guide
- `fixtext_audit_rules_dac_modification_chown` was defined twice
- defaults sometimes were for wrong type
- some case clearly optional params were not implemented as such
- `ocil_firefox_config` had bad jinja
- remove unused macros, either never used, or usage replaced with something else, is there some jinja macro stability guarantee and no macro can be removed? Because there was multiple times some kind of macro rename and old one was then kept around. That does not work for long run and some of the removed macros are really complex and could easily catch some maintenance work to be wasted. Each commit should have hint where usage was removed.

I wonder if it would be feasible to remove all unused macros?
- `check_file_permissions`: similar functionality existing and there is potential to replace the content with the macro `file_permissions_cron_daily` for example
- `describe_iptables_block`: `describe_iptables_allow` existing, so keep it now

Macro usage can be seen like:
```
for a in $(grep -Eh '\{\{%-? macro ' shared/macros/* |sed -E 's/\(.*//;s/.* //'|sort -u);do echo "$(git grep "$a"|wc -l) $a";done|sort -n
```

#### Rationale:

I removed my tool from here as it is not up to par as is, code is available now at: https://github.com/maage/scap-security-guide/tree/doc-3

#### Review Hints:

Especially look style guide and removed macros.

### CODE REVIEW

**Comment by maage** [comment]
Realized that some macros were only mentioned in their own comment block.

File: shared/macros/01-general.jinja [hunk 1]
```diff
@@ -246,12 +253,18 @@ Therefore, you need to use a tool that can query the OCP API, retrieve the follo
...
-{{%- macro describe_firewalld_allow(port, proto, service) %}}
-    {{% if port %}}
+{{%- macro describe_firewalld_allow(port=none, proto=none, service=none) %}}
+    {{%- if ((port is none) and (proto is not none)) or ((port is not none) and (proto is none)) %}}
+    {{{- raise("describe_firewalld_allow: if you have proto/port you need respectively port/proto") }}}
+    {{%- endif %}}
+    {{%- if (service is none) and (port is none) %}}
+    {{{- raise("describe_firewalld_allow: define either service, or port+proto") }}}
+    {{%- endif %}}
+    {{% if port is not none %}}
```
Comment (matejak) [inline shared/macros/01-general.jinja:263]: This forest of ifs is quite complex, and we have just a couple of proto/service and one proto/port. Would it work for you to split the macro instead? Not only it has to figure out how it was called at the beginning, but the effective body is also a one big if/else.

File: shared/macros/10-ocil.jinja [hunk 2]
```diff
@@ -639,13 +639,14 @@ ocil_clause: '"{{{ part }}} is not a mountpoint" is returned'
 
 {{# Firewalld macros #}}
 
-{{%- macro _firewalld_check(access_action, port, proto, service) %}}
-    To determine if <code>firewalld</code> is configured to {{{ access_action }}} to <code>{{{ service }}}</code>
+{{%- macro _firewalld_check(access_action, port=none, proto=none, service=none) %}}
+    To determine if <code>firewalld</code> is configured to {{{ access_action }}}
+    {{% if port is not none %}}
```
Comment (matejak) [inline shared/macros/10-ocil.jinja:644]: This is a nice fix, but let's be a bit stricter - based on the usage, it looks like that all arguments are passed, so let's make them required. The uncertainty of what behaviors should be suppoted, lack of tests and the messy nature of Jinja make those macros like small puzzles, while they should be mainly tools.

**Comment by matejak** [comment]
Thank you for the PR, i appreciate the effort into fixing this part of the project.
The good thing about the CI is that it will fail when a macro that is used somewhere disappears. The dilemma whether to keep a potentially useful macro around or not doesn't have an easy conclusion, but I also lean in favor of removing, as they remain in git history, and perhaps we will be able to ask AI about such macros in the near future.
So generally, I suggest to make our life simpler, if a macro can do a lot of things, but they are not used, let's remove the unused functionality, let's get rid of optional arguments, and perhaps let's split macros if they are already split inside by an if condition. Various nested or chained if/else Jinja blocks make working with macros painful.

File: shared/macros/01-general.jinja (hunk 1 above)
Comment (maage) [reply shared/macros/01-general.jinja:263]: Ok, it simplifies.

File: shared/macros/10-ocil.jinja (hunk 2 above)
Comment (maage) [reply shared/macros/10-ocil.jinja:644]: Ok

**Comment by maage** [comment]
Rebased and handled change requests.

One aspect not used as much as I'd like is to check jinja parameters for suitability. That makes macros somewhat tedious. But if macro is not simplest, then it probably is better to catch issues sooner than later. 

It should be possible to "monkeypatch" macro definitions to do basic type checking, https://github.com/maage/scap-security-guide/blob/doc-3/utils/macro-doc-check.py has code to generate fragments. `ssg.jinja.update_substitutions_dict` just needs to be modified to modify a jinja macro template file to add fragments to macro definition lines. To not to modify line numbers. But that is left to some later date. I guess this kind of functionality should be implemented as jinja plugin or something like that as it would be usabe in any other jinja project too.

**Review by matejak — APPROVED** [review]
Sorry for the delay!
The PR looks good, it just developed a conflict regarding removal of the duplicate macros, so just kick the `fixtext_audit_rules_dac_modification_chown` commit out, and you should be fine.

@jan-cerny Could you please merge as soon as the last obstacle dissolves?

**Comment by matejak** [comment]
> One aspect not used as much as I'd like is to check jinja parameters for suitability. That makes macros somewhat tedious. But if macro is not simplest, then it probably is better to catch issues sooner than later.
> 
> It should be possible to "monkeypatch" macro definitions to do basic type checking, https://github.com/maage/scap-security-guide/blob/doc-3/utils/macro-doc-check.py has code to generate fragments. `ssg.jinja.update_substitutions_dict` just needs to be modified to modify a jinja macro template file to add fragments to macro definition lines. To not to modify line numbers. But that is left to some later date. I guess this kind of functionality should be implemented as jinja plugin or something like that as it would be usabe in any other jinja project too.

I would be careful here. The reason why we introduced Jinja was to provide a declarative-like input interface that would generate everything that was needed. Now, we ended up with a monstrous code that is difficult understand, to extend, and to test. You are right, and I believe that the inner complexity and opacity spreads the mess in all directions.
I think that we could aim at larger changes - e.g. use jinja to populate a dictionary, but use a Python class passed to Jinja to assemble the remediation.

**Comment by jan-cerny** [comment]
@matejak the obstacle hasn't dissolved and moreover a merge conflict now needs to be solved

**Comment by marcusburghardt** [comment]
This PR was already reviewed and approved. @matejak , maybe you can rebase it and merge?

**Comment by matejak** [comment]
Oh sorry, I haven't noticed. Anyway, there is nothing to wait for, merging.

**Pull request merged**